clippy_lints/src/manual_abs_diff.rs

```diff
@@ -0,0 +1,168 @@
+    ///
+    /// ### Examples
+    /// ```no_run
+    /// # let (a, b) = (5_usize, 3_usize);
+    /// if a > b {
+    ///     a - b
+    /// } else {
+    ///     b - a
+    /// }
+    /// # ;
+    /// ```
+    /// Use instead:
+    /// ```no_run
+    /// # let (a, b) = (5_usize, 3_usize);
+    /// a.abs_diff(b)
+    /// # ;
+    /// ```
+    #[clippy::version = "1.86.0"]
+    pub MANUAL_ABS_DIFF,
+    complexity,
+    "using an if-else pattern instead of `abs_diff`"
+}
+
+impl_lint_pass!(ManualAbsDiff => [MANUAL_ABS_DIFF]);
+
+pub struct ManualAbsDiff {
+    msrv: Msrv,
+}
+
+impl ManualAbsDiff {
+    pub fn new(conf: &'static Conf) -> Self {
+        Self { msrv: conf.msrv }
+    }
+}
+
+#[derive(Debug)]
+struct Suggestion<'tcx> {
+    params: Input<'tcx>,
+    span: Span,
+}
+
+#[derive(Debug)]
+struct Input<'tcx> {
+    a: &'tcx Expr<'tcx>,
+    b: &'tcx Expr<'tcx>,
+}
+
+impl<'tcx> LateLintPass<'tcx> for ManualAbsDiff {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>) {
+        if !expr.span.from_expansion()
+            && let Some(suggestion) = self.is_manual_abs_diff_pattern(cx, expr)
+        {
+            emit_suggestion(cx, &suggestion);
+        }
+    }
+}
+
+impl ManualAbsDiff {
+    /// Matches a pattern such as `if a > b { a - b } else { b - a }`
+    fn is_manual_abs_diff_pattern<'tcx>(
+        &self,
+        cx: &LateContext<'tcx>,
+        expr: &'tcx Expr<'tcx>,
+    ) -> Option<Suggestion<'tcx>> {
+        if let Some(If {
+            cond,
+            then,
+            r#else: Some(r#else),
+        }) = If::hir(expr)
+            && let ExprKind::Binary(
+                Spanned {
+                    node: BinOpKind::Gt | BinOpKind::Ge,
+                    ..
+                },
+                a,
+                b,
+            )
+            | ExprKind::Binary(
+                Spanned {
+                    node: BinOpKind::Lt | BinOpKind::Le,
+                    ..
+                },
+                b,
+                a,
+            ) = cond.kind
+            && self.is_ty_applicable(cx, a)
+            && self.is_ty_applicable(cx, b)
+            && Self::is_sub_expr(cx, then, a, b)
+            && Self::is_sub_expr(cx, r#else, b, a)
+        {
+            Some(Suggestion {
+                params: Input { a, b },
+                span: expr.span,
+            })
+        } else {
+            None
+        }
+    }
+
+    /// Returns `true` if `expr` is a type that this lint can be applied to (currently, any
+    /// primitive unsigned int, or a `Duration`)
+    fn is_ty_applicable(&self, cx: &LateContext<'_>, expr: &Expr<'_>) -> bool {
+        let expr_ty = cx.typeck_results().expr_ty(expr).peel_refs();
+        (matches!(expr_ty.kind(), ty::Uint(_)) && self.msrv.meets(cx, msrvs::ABS_DIFF))
+            || (is_type_diagnostic_item(cx, expr_ty, sym::Duration) && self.msrv.meets(cx, msrvs::DURATION_ABS_DIFF))
+    }
```

Comment:
Rather than checking each type individually, you can first make sure that they are identical, then check for their eligibility for the lint:

```suggestion
    fn are_ty_eligible(&self, cx: &LateContext<'_>, a: &Expr<'_>, b: &Expr<'_>) -> bool {
        let a_ty = cx.typeck_results().expr_ty(a).peel_refs();
        a_ty == cx.typeck_results().expr_ty(b).peel_refs()
            && ((matches!(a_ty.kind(), ty::Uint(_)) && self.msrv.meets(cx, msrvs::ABS_DIFF))
                || (is_type_diagnostic_item(cx, a_ty, sym::Duration) && self.msrv.meets(cx, msrvs::DURATION_ABS_DIFF)))
    }
```